src/main/java/com/epam/processor/DataProcessor.java

```diff
@@ -92,32 +157,80 @@ public RoadAccident getAccidentByIndex(String index){
+        return roadAccidentList.stream()
+                .filter(roadAccident -> isGeoPositionInsideBoundaries(
+                        roadAccident.getLatitude(),
+                        roadAccident.getLongitude(),
+                        minLongitude, maxLongitude, minLatitude, maxLatitude))
+                .collect(toList());
+    }
+
+    private boolean isGeoPositionInsideBoundaries(float lat, float lon, float minLongitude, float maxLongitude, float minLatitude, float maxLatitude) {
+        return lat > minLatitude && lat < maxLatitude && lon > minLongitude && lon < maxLongitude;
     }
 
     /**
      * find the weather conditions which caused max number of incidents
      * @return
      */
-    public List<String> getTopThreeWeatherCondition(){
-        return null;
+    public List<String> getTopThreeWeatherCondition(List<RoadAccident> roadAccidentList){
+        Map<String, Long> weatherStatistics = roadAccidentList.stream()
+                .collect(groupingBy(RoadAccident::getWeatherConditions, counting()));
+
+        return weatherStatistics.entrySet().stream()
+                .sorted((o1, o2) -> o2.getValue().compareTo(o1.getValue()))
+                .map(stringLongEntry -> stringLongEntry.getKey())
+                .limit(3)
+                .collect(toList());
     }
 
     /**
      * count incidents by road surface conditions
      * @return
      */
-    public Map<String, Long> getCountByRoadSurfaceCondition(){
-        return null;
+    public Map<String, Long> getCountByRoadSurfaceCondition(List<RoadAccident> roadAccidentList){
+        return roadAccidentList.stream()
+                .collect(groupingBy(RoadAccident::getRoadSurfaceConditions, counting()));
     }
 
     /**
      * To match streaming operations result, return type is a java collection instead of multimap
      * @return
      */
-    public Map<String, List<String>> getAccidentIdsGroupedByAuthority(){
-        return null;
+    public Map<String, List<String>> getAccidentIdsGroupedByAuthority(List<RoadAccident> roadAccidentList){
+        return roadAccidentList.stream()
+                .collect(groupingBy(
+                        RoadAccident::getDistrictAuthority,
+                        HashMap::new,
+                        mapping(RoadAccident::getAccidentId, toList())
+                ));
+    }
+
+    public static void enrichWithTimeOfDay(List<RoadAccident> accidents) {
+        accidents.stream()
+                .forEach(roadAccident -> roadAccident.setTimeOfDay(timeOfDayFromTime(roadAccident.getTime())));
+    }
+
+    private static TimeOfDay timeOfDayFromTime(LocalTime time) {
+        for (TimeOfDay part : TimeOfDay.values()) {
+            if (!time.isAfter(part.getEnd()) && !time.isBefore(part.getStart())) {
+                return part;
+            }
+        }
+
+        throw new RuntimeException("Could not define day part for time provided: " + time.toString());
```

Comment:
Agree. I put this exception for debugging reasons as in production there should be no situation when we can't decide where to which day part certain point of time is related. Remove it.